File: internal/hooks/precompact.go
// pre-compact hook handler.
// Ports runPreCompact from scripts/hooks/pre-compact.js.
// Always exits 0 and emits {"continue":true}. Uses internal/store + internal/yamllite.
package hooks

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/mretamozo-hiberuscom/ospec-workflow/internal/store"
	"github.com/mretamozo-hiberuscom/ospec-workflow/internal/yamllite"
)

func init() {
	Register(&preCompactHandler{})
}

type preCompactHandler struct{}

func (h *preCompactHandler) Name() string { return "pre-compact" }

// phaseRanks mirrors PHASE_RANKS in pre-compact.js.
var phaseRanks = map[string]int{
	"explore":    1,
	"exploration": 1,
	"propose":    2,
	"proposal":   2,
	"spec":       3,
	"specs":      3,
	"design":     3,
	"tasks":      4,
	"apply":      5,
	"verify":     6,
	"archive":    7,
}

// artifactCandidates mirrors ARTIFACT_CANDIDATES in pre-compact.js.
type artifactCandidate struct {
	relativePath string
	rank         int
}

var artifactCandidates = []artifactCandidate{
	{"exploration.md", 1},
	{"proposal-lite.md", 2},
	{"proposal.md", 2},
	{"design.md", 3},
	{"tasks.md", 4},
	{"apply-progress.md", 5},
	{"verify-report.md", 6},
	{"archive-report.md", 7},
}

func (h *preCompactHandler) Run(stdin []byte) ([]byte, int) {
	var input struct {
		Cwd string `json:"cwd"`
	}
	if err := json.Unmarshal(stdin, &input); err != nil {
		return continueWithError(fmt.Sprintf("PreCompact could not persist the session summary: %s", err.Error())), 0
	}

	if err := runPreCompact(input.Cwd); err != nil {
		return continueWithError(fmt.Sprintf("PreCompact could not persist the session summary: %s", err.Error())), 0
	}

	b, _ := json.Marshal(map[string]bool{"continue": true})
	return b, 0
}

func runPreCompact(cwd string) error {
	workspace := resolveCwd(cwd)
	s := store.NewStore(workspace)

	changes, err := s.FindActiveChanges()
	if err != nil {
		return err
	}
	if len(changes) == 0 {
		return nil // no active change → silently skip
	}
	activeChange := changes[0]

	changeName := yamllite.ExtractFirstScalar(activeChange.Content, [][]string{
		{"change", "name"},
	})
	if changeName == "" {
		changeName = activeChange.DirectoryName
	}

	currentPhase := yamllite.ExtractFirstScalar(activeChange.Content, [][]string{
		{"change", "current_phase"},
		{"current_phase"},
		{"phase"},
	})

	lastArtifact, err := inferLastCompletedArtifact(workspace, activeChange, currentPhase)
	if err != nil {
		return err
	}

	approvals := formatApprovalsPC(activeChange.Content)
	blockers := formatBlockersPC(activeChange.Content)
	nextRec := yamllite.ExtractFirstScalar(activeChange.Content, [][]string{{"next_recommended"}})
	nextAction := yamllite.FormatNextAction(nextRec, changeName)

	summary := renderSummaryPC(renderSummaryPCArgs{
		changeName:           changeName,
		currentPhase:         currentPhase,
		lastCompletedArtifact: lastArtifact,
		blockers:             blockers,
		approvals:            approvals,
		nextAction:           nextAction,
	})

	_, err = s.WriteSessionSummary(activeChange.DirectoryName, summary)
	return err
}

// ── last-artifact inference ────────────────────────────────────────────────────

func normalizePhase(v string) string {
	v = strings.TrimPrefix(strings.ToLower(strings.TrimSpace(v)), "sdd-")
	return v
}

func toPortablePathPC(p string) string {
	return filepath.ToSlash(p)
}

func pathIsFilePC(p string) bool {
	info, err := os.Stat(p)
	return err == nil && info.Mode().IsRegular()
}

func collectSpecArtifactsPC(changeDir string) []artifactCandidate {
	specsRoot := filepath.Join(changeDir, "specs")
	var out []artifactCandidate

	_ = filepath.Walk(specsRoot, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if !info.IsDir() && info.Name() == "spec.md" {
			rel := toPortablePathPC(path[len(changeDir)+1:]) // strip changeDir prefix + sep
			if len(path) > len(changeDir) {
				rel = toPortablePathPC(strings.TrimPrefix(path, changeDir+string(filepath.Separator)))
			}
			out = append(out, artifactCandidate{rel, 3})
		}
		return nil
	})

	sort.Slice(out, func(i, j int) bool { return out[i].relativePath < out[j].relativePath })
	return out
}

func inferLastCompletedArtifact(workspace string, activeChange *store.ActiveChange, phase string) (string, error) {
	// Check for explicit override.
	explicit := yamllite.ExtractFirstScalar(activeChange.Content, [][]string{
		{"runtime", "last_completed_artifact"},
		{"last_completed_artifact"},
	})
	if explicit != "" {
		return toPortablePathPC(explicit), nil
	}

	currentRank := phaseRanks[normalizePhase(phase)]
	if currentRank == 0 {
		currentRank = 1<<31 - 1 // infinity
	}

	candidates := append(append([]artifactCandidate{}, artifactCandidates...), collectSpecArtifactsPC(activeChange.ChangeDirectory)...)

	var existing []artifactCandidate
	for _, c := range candidates {
		if c.rank < currentRank && pathIsFilePC(filepath.Join(activeChange.ChangeDirectory, filepath.FromSlash(c.relativePath))) {
			existing = append(existing, c)
		}
	}

	if len(existing) == 0 {
		return "None", nil
	}

	// Sort: highest rank first, then reverse-alpha.
	sort.Slice(existing, func(i, j int) bool {
		if existing[i].rank != existing[j].rank {
			return existing[i].rank > existing[j].rank
		}
		return existing[i].relativePath > existing[j].relativePath
	})

	best := existing[0]
	abs := filepath.Join(activeChange.ChangeDirectory, filepath.FromSlash(best.relativePath))
	rel, err := filepath.Rel(workspace, abs)
	if err != nil {
		return toPortablePathPC(abs), nil
	}
	return toPortablePathPC(rel), nil
}

// ── rendering ─────────────────────────────────────────────────────────────────

type renderSummaryPCArgs struct {
	changeName            string
	currentPhase          string
	lastCompletedArtifact string
	blockers              []string
	approvals             []string
	nextAction            string
}

func renderListPC(items []string) string {
	if len(items) == 0 {
		return "- None"
	}
	var sb strings.Builder
	for i, v := range items {
		if i > 0 {
			sb.WriteByte('\n')
		}
		sb.WriteString("- ")
		sb.WriteString(v)
	}
	return sb.String()
}

func renderSummaryPC(args renderSummaryPCArgs) string {
	phase := args.currentPhase
	if phase == "" {
		phase = "unknown"
	}
	parts := []string{
		"# Session Summary",
		"",
		"## Active change",
		"`" + args.changeName + "`",
		"",
		"## Current phase",
		"`" + phase + "`",
		"",
		"## Last completed artifact",
		"`" + args.lastCompletedArtifact + "`",
		"",
		"## Blocking decisions",
		renderListPC(args.blockers),
		"",
		"## Approvals",
		renderListPC(args.approvals),
		"",
		"## Next recommended action",
		args.nextAction,
		"",
	}
	return strings.Join(parts, "\n")
}

// formatBlockersPC mirrors formatBlockers in pre-compact.js.
func formatBlockersPC(content string) []string {
	for _, sectionName := range []string{"blocking_questions", "blockers"} {
		items := yamllite.ExtractListSection(content, sectionName)
		if len(items) > 0 {
			var out []string
			for _, item := range items {
				if item.Value != "" {
					out = append(out, item.Value)
				} else if item.Fields != nil {
					// Pick first of: question, message, reason, id, or JSON.
					for _, key := range []string{"question", "message", "reason", "id"} {
						if v, ok := item.Fields[key]; ok && v != "" {
							out = append(out, v)
							break
						}
					}
				}
			}
			if len(out) > 0 {
				return out
			}
		}
		// Scalar fallback.
		scalar := yamllite.ExtractFirstScalar(content, [][]string{{sectionName}})
		if scalar != "" && strings.ToLower(scalar) != "none" {
			return []string{scalar}
		}
	}
	return nil
}

// formatApprovalsPC mirrors formatApprovals in pre-compact.js.
func formatApprovalsPC(content string) []string {
	items := yamllite.ExtractListSection(content, "approvals")
	var out []string
	for _, item := range items {
		if item.Value != "" {
			out = append(out, item.Value)
		} else if item.Fields != nil {
			gate := item.Fields["gate"]
			if gate == "" {
				gate = item.Fields["id"]
			}
			if gate == "" {
				gate = "approval"
			}
			decision := item.Fields["decision"]
			if decision == "" {
				decision = item.Fields["status"]
			}
			if decision == "" {
				decision = "recorded"
			}
			out = append(out, gate+": "+decision)
		}
	}
	return out
}

// resolveCwd resolves the workspace from the input cwd field (or returns ".").
func resolveCwd(cwd string) string {
	ws := strings.TrimSpace(cwd)
	if ws == "" {
		return "."
	}
	return filepath.Clean(ws)
}

// continueWithError returns a {"continue":true,"systemMessage":"..."} payload.
func continueWithError(msg string) []byte {
	b, _ := json.Marshal(map[string]any{
		"continue":      true,
		"systemMessage": msg,
	})
	return b
}

File: plugins/ospec-workflow/scripts/hooks/pre-compact.js
#!/usr/bin/env node

"use strict";

const fs = require("node:fs/promises");
const path = require("node:path");
const {
  createArtifactStoreFromConfig,
} = require("../lib/artifact-store.js");
const { resolveWorkspaceCwd } = require("../lib/pathsafe.js");

const PHASE_RANKS = new Map([
  ["explore", 1],
  ["exploration", 1],
  ["propose", 2],
  ["proposal", 2],
  ["spec", 3],
  ["specs", 3],
  ["design", 3],
  ["tasks", 4],
  ["apply", 5],
  ["verify", 6],
  ["archive", 7],
]);

const ARTIFACT_CANDIDATES = [
  { relativePath: "exploration.md", rank: 1 },
  { relativePath: "proposal-lite.md", rank: 2 },
  { relativePath: "proposal.md", rank: 2 },
  { relativePath: "design.md", rank: 3 },
  { relativePath: "tasks.md", rank: 4 },
  { relativePath: "apply-progress.md", rank: 5 },
  { relativePath: "verify-report.md", rank: 6 },
  { relativePath: "archive-report.md", rank: 7 },
];

function compareStrings(left, right) {
  if (left < right) {
    return -1;
  }

  return left > right ? 1 : 0;
}

function toPortablePath(filePath) {
  return filePath.split(path.sep).join("/");
}

function parseScalar(value) {
  const trimmed = value.trim();

  if (!trimmed) {
    return "";
  }

  const quoted = trimmed.match(/^(["'])([\s\S]*)\1$/);

  if (quoted) {
    return quoted[2];
  }

  return trimmed.replace(/\s+#.*$/, "").trim();
}

function parseYamlLines(content) {
  return content
    .split(/\r?\n/)
    .map((raw) => ({
      indent: raw.match(/^\s*/)[0].length,
      content: raw.trim(),
    }))
    .filter((line) => line.content && !line.content.startsWith("#"));
}

function extractScalarAtPath(content, expectedPath) {
  const stack = [];

  for (const line of parseYamlLines(content)) {
    if (line.content.startsWith("- ")) {
      continue;
    }

    const match = line.content.match(/^([^:]+):(?:\s*(.*))?$/);

    if (!match) {
      continue;
    }

    while (stack.length && stack[stack.length - 1].indent >= line.indent) {
      stack.pop();
    }

    const key = match[1].trim();
    const value = match[2] || "";
    const currentPath = [...stack.map((entry) => entry.key), key];

    if (
      currentPath.length === expectedPath.length &&
      currentPath.every((part, index) => part === expectedPath[index]) &&
      value
    ) {
      return parseScalar(value);
    }

    if (!value) {
      stack.push({ indent: line.indent, key });
    }
  }

  return "";
}

function extractFirstScalar(content, paths) {
  for (const expectedPath of paths) {
    const value = extractScalarAtPath(content, expectedPath);

    if (value) {
      return value;
    }
  }

  return "";
}

function extractTopLevelSection(content, sectionName) {
  const lines = content.split(/\r?\n/);
  const section = [];
  let sectionIndent = -1;
  let collecting = false;

  for (const raw of lines) {
    const trimmed = raw.trim();
    const indent = raw.match(/^\s*/)[0].length;

    if (!collecting) {
      if (
        indent === 0 &&
        new RegExp(`^${sectionName}:\\s*(?:#.*)?$`).test(trimmed)
      ) {
        collecting = true;
        sectionIndent = indent;
      }
      continue;
    }

    if (trimmed && indent <= sectionIndent) {
      break;
    }

    section.push(raw);
  }

  return section;
}

function extractListSection(content, sectionName) {
  const section = extractTopLevelSection(content, sectionName);
  const values = [];
  let current = null;
  let itemIndent = null;

  for (const raw of section) {
    const trimmed = raw.trim();
    const indent = raw.match(/^\s*/)[0].length;

    if (!trimmed || trimmed.startsWith("#")) {
      continue;
    }

    if (trimmed.startsWith("- ")) {
      if (itemIndent === null) {
        itemIndent = indent;
      }

      if (indent !== itemIndent) {
        continue;
      }

      const item = trimmed.slice(2).trim();
      const field = item.match(/^([^:]+):\s*(.*)$/);

      if (field) {
        current = { [field[1].trim()]: parseScalar(field[2]) };
        values.push(current);
      } else {
        current = null;
        values.push(parseScalar(item));
      }
      continue;
    }

    if (
      current &&
      typeof current === "object" &&
      itemIndent !== null &&
      indent === itemIndent + 2
    ) {
      const field = trimmed.match(/^([^:]+):\s*(.*)$/);

      if (field) {
        current[field[1].trim()] = parseScalar(field[2]);
      }
    }
  }

  return values;
}

function normalizePhase(value) {
  return String(value || "")
    .replace(/^sdd-/, "")
    .trim()
    .toLowerCase();
}

async function findActiveChange(workspace, mode) {
  const store = await createArtifactStoreFromConfig({ mode, workspace });

  return (await store.findActiveChanges())[0] || null;
}

async function collectSpecArtifacts(changeDirectory) {
  const specsRoot = path.join(changeDirectory, "specs");
  const artifacts = [];

  async function visit(directory) {
    let entries;

    try {
      entries = await fs.readdir(directory, { withFileTypes: true });
    } catch (error) {
      if (error.code === "ENOENT") {
        return;
      }

      throw error;
    }

    for (const entry of entries) {
      const absolutePath = path.join(directory, entry.name);

      if (entry.isDirectory()) {
        await visit(absolutePath);
      } else if (entry.isFile() && entry.name === "spec.md") {
        artifacts.push({
          relativePath: toPortablePath(
            path.relative(changeDirectory, absolutePath),
          ),
          rank: 3,
        });
      }
    }
  }

  await visit(specsRoot);
  return artifacts.sort((left, right) =>
    compareStrings(left.relativePath, right.relativePath),
  );
}

async function pathIsFile(filePath) {
  try {
    return (await fs.stat(filePath)).isFile();
  } catch (error) {
    if (error.code === "ENOENT") {
      return false;
    }

    throw error;
  }
}

async function inferLastCompletedArtifact(workspace, activeChange, phase) {
  const explicitPath = extractFirstScalar(activeChange.content, [
    ["runtime", "last_completed_artifact"],
    ["last_completed_artifact"],
  ]);

  if (explicitPath) {
    return toPortablePath(explicitPath);
  }

  const currentRank = PHASE_RANKS.get(normalizePhase(phase)) || Infinity;
  const candidates = [
    ...ARTIFACT_CANDIDATES,
    ...(await collectSpecArtifacts(activeChange.changeDirectory)),
  ];
  const existing = [];

  for (const candidate of candidates) {
    if (
      candidate.rank < currentRank &&
      (await pathIsFile(
        path.join(activeChange.changeDirectory, candidate.relativePath),
      ))
    ) {
      existing.push(candidate);
    }
  }

  existing.sort(
    (left, right) =>
      right.rank - left.rank ||
      compareStrings(right.relativePath, left.relativePath),
  );

  if (!existing.length) {
    return "None";
  }

  return toPortablePath(
    path.relative(
      workspace,
      path.join(activeChange.changeDirectory, existing[0].relativePath),
    ),
  );
}

function formatBlockers(content) {
  for (const sectionName of ["blocking_questions", "blockers"]) {
    const values = extractListSection(content, sectionName);

    if (values.length) {
      return values.map((value) => {
        if (typeof value === "string") {
          return value;
        }

        return (
          value.question ||
          value.message ||
          value.reason ||
          value.id ||
          JSON.stringify(value)
        );
      });
    }

    const scalar = extractFirstScalar(content, [[sectionName]]);

    if (scalar && scalar.toLowerCase() !== "none") {
      return [scalar];
    }
  }

  return [];
}

function formatApprovals(content) {
  return extractListSection(content, "approvals").map((approval) => {
    if (typeof approval === "string") {
      return approval;
    }

    const gate = approval.gate || approval.id || "approval";
    const decision = approval.decision || approval.status || "recorded";

    return `${gate}: ${decision}`;
  });
}

function formatNextAction(value, changeName) {
  const next = String(value || "").trim();

  if (!next) {
    return `Run \`sdd-continue ${changeName}\`.`;
  }

  if (next.toLowerCase() === "none") {
    return "None.";
  }

  if (/^\/?sdd-[a-z-]+$/i.test(next)) {
    return `Run \`${next.replace(/^\//, "")} ${changeName}\`.`;
  }

  return next.endsWith(".") ? next : `${next}.`;
}

function renderList(values) {
  return values.length ? values.map((value) => `- ${value}`).join("\n") : "- None";
}

function renderSummary({
  approvals,
  blockers,
  changeName,
  currentPhase,
  lastCompletedArtifact,
  nextAction,
}) {
  return [
    "# Session Summary",
    "",
    "## Active change",
    `\`${changeName}\``,
    "",
    "## Current phase",
    `\`${currentPhase || "unknown"}\``,
    "",
    "## Last completed artifact",
    `\`${lastCompletedArtifact}\``,
    "",
    "## Blocking decisions",
    renderList(blockers),
    "",
    "## Approvals",
    renderList(approvals),
    "",
    "## Next recommended action",
    nextAction,
    "",
  ].join("\n");
}

async function runPreCompact({
  input = {},
  fallbackCwd = process.cwd(),
  mode,
} = {}) {
  const workspace = resolveWorkspaceCwd(input.cwd, fallbackCwd);
  const store = await createArtifactStoreFromConfig({ mode, workspace });
  const activeChange = (await store.findActiveChanges())[0] || null;

  if (!activeChange) {
    return { status: "skipped", reason: "no-active-change" };
  }

  const changeName =
    extractFirstScalar(activeChange.content, [["change", "name"]]) ||
    activeChange.directoryName;
  const currentPhase = extractFirstScalar(activeChange.content, [
    ["change", "current_phase"],
    ["current_phase"],
    ["phase"],
  ]);
  const lastCompletedArtifact = await inferLastCompletedArtifact(
    workspace,
    activeChange,
    currentPhase,
  );
  const summary = renderSummary({
    approvals: formatApprovals(activeChange.content),
    blockers: formatBlockers(activeChange.content),
    changeName,
    currentPhase,
    lastCompletedArtifact,
    nextAction: formatNextAction(
      extractFirstScalar(activeChange.content, [["next_recommended"]]),
      changeName,
    ),
  });
  const writeResult = await store.writeSessionSummary(
    activeChange.directoryName,
    summary,
  );

  return {
    status: writeResult.status,
    change: changeName,
    path: writeResult.path,
  };
}

async function readJsonInput(stream = process.stdin) {
  const chunks = [];

  for await (const chunk of stream) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
  }

  const input = Buffer.concat(chunks).toString("utf8").trim();
  return input ? JSON.parse(input) : {};
}

async function main() {
  try {
    await runPreCompact({ input: await readJsonInput() });
    process.stdout.write('{"continue":true}\n');
  } catch (error) {
    process.stdout.write(
      `${JSON.stringify({
        continue: true,
        systemMessage: `PreCompact could not persist the session summary: ${error.message}`,
      })}\n`,
    );
  }
}

if (require.main === module) {
  void main();
}

module.exports = {
  extractFirstScalar,
  extractListSection,
  findActiveChange,
  formatNextAction,
  renderSummary,
  runPreCompact,
};
